File: swarm-refactored/src/core/redis-bus.ts
import Redis from 'ioredis';
import type {
  A2AMessage,
  DefenseAnalytics,
  ExploitResult,
  ReconResult,
  DiscoveredToken,
  BlueTeamFinding,
} from '../types/index.js';

const STREAMS = {
  A2A_MESSAGES: 'a2a_messages',
  RED_TEAM_EVENTS: 'red_team_events',
  DEFENSE_ANALYTICS: 'defense_analytics',
};

const KEY_PREFIX = 'redteam';

export class RedisBus {
  private _client: Redis | null = null;
  private subscriber: Redis | null = null;
  private url: string;

  constructor(url: string = 'redis://localhost:6379/0') {
    this.url = url;
  }

  getClient(): Redis {
    if (!this._client) {
      throw new Error('Redis not connected');
    }
    return this._client;
  }

  async connect(): Promise<void> {
    const url = process.env.REDIS_URL || this.url;
    this._client = new Redis(url, {
      maxRetriesPerRequest: 3,
      lazyConnect: true,
    });

    this.subscriber = new Redis(url, {
      maxRetriesPerRequest: 3,
      lazyConnect: true,
    });

    this._client.on('error', () => {});
    this.subscriber.on('error', () => {});

    try {
      await this._client.connect();
      await this._client.ping();
    } catch {
      this._client = null;
      this.subscriber = null;
      throw new Error('Redis connection failed');
    }
  }

  async disconnect(): Promise<void> {
    if (this.subscriber) {
      await this.subscriber.quit();
      this.subscriber = null;
    }
    if (this._client) {
      await this._client.quit();
      this._client = null;
    }
  }

  async publish(stream: string, message: Record<string, string>): Promise<void> {
    if (!this._client) return;
    await this._client.xadd(stream, '*', ...this.flattenObject(message));
  }

  async consume(
    stream: string,
    group: string,
    consumer: string,
    count: number = 10
  ): Promise<Array<{ id: string; message: Record<string, string> }>> {
    if (!this._client) return [];

    try {
      const result = await this._client.xreadgroup(
        'GROUP',
        group,
        consumer,
        'COUNT',
        count,
        'BLOCK',
        5000,
        'STREAMS',
        stream,
        '>'
      ) as unknown as Array<[string, Array<[string, string[]]>]> | null;

      if (!result || !result[0]) return [];

      return result[0][1].map((item: [string, string[]]) => ({
        id: item[0],
        message: this.unflattenObject(item[1]),
      }));
    } catch {
      return [];
    }
  }

  async ack(stream: string, group: string, id: string): Promise<void> {
    if (!this._client) return;
    await this._client.xack(stream, group, id);
  }

  async blackboard_write(
    missionId: string,
    key: string,
    value: unknown
  ): Promise<void> {
    if (!this._client) return;
    const fullKey = `${KEY_PREFIX}:blackboard:${missionId}`;
    await this._client.hset(fullKey, key, JSON.stringify(value));
  }

  async blackboard_read<T = unknown>(missionId: string, key: string): Promise<T | null> {
    if (!this._client) return null;
    const fullKey = `${KEY_PREFIX}:blackboard:${missionId}`;
    const value = await this._client.hget(fullKey, key);
    return value ? JSON.parse(value) : null;
  }

  async blackboard_read_all(missionId: string): Promise<Record<string, unknown>> {
    if (!this._client) return {};
    const fullKey = `${KEY_PREFIX}:blackboard:${missionId}`;
    const data = await this._client.hgetall(fullKey);
    const result: Record<string, unknown> = {};
    for (const [key, value] of Object.entries(data)) {
      result[key] = JSON.parse(value);
    }
    return result;
  }

  async blackboard_clear(missionId: string): Promise<void> {
    if (!this._client) return;
    const fullKey = `${KEY_PREFIX}:blackboard:${missionId}`;
    await this._client.del(fullKey);
  }

  async consume_defense_analytics(
    group: string = 'swarm-consumers',
    consumer: string = 'swarm',
    count: number = 10
  ): Promise<Array<{ id: string; analytics: DefenseAnalytics }>> {
    if (!this._client) return [];

    try {
      await this._client.xgroup(
        'CREATE',
        STREAMS.DEFENSE_ANALYTICS,
        group,
        'MKSTREAM',
        'ENTRIESREAD',
        '0'
      );
    } catch {
      // Group may already exist
    }

    const messages = await this.consume(
      STREAMS.DEFENSE_ANALYTICS,
      group,
      consumer,
      count
    );

    return messages
      .map(({ id, message }) => ({
        id,
        analytics: {
          severity: message.severity as DefenseAnalytics['severity'],
          vulnerability_type: message.vulnerability_type,
          description: message.description,
          blocked_payload: message.blocked_payload,
          detected_signature: message.detected_signature,
          endpoint: message.endpoint,
          target: message.target,
          mission_id: message.mission_id,
          source: message.source as DefenseAnalytics['source'],
          agent: message.agent,
          timestamp: message.timestamp,
        },
      }))
      .filter((m) => m.analytics.mission_id !== undefined);
  }

  async get_latest_defense_intel(
    missionId?: string
  ): Promise<DefenseAnalytics[]> {
    if (!this._client) return [];

    const messages = await this.consume_defense_analytics();

    const filtered = messages.filter(
      (m) => !missionId || m.analytics.mission_id === missionId
    );

    return filtered.map((m) => m.analytics);
  }

  async findings_store(
    missionId: string,
    category: string,
    key: string,
    value: string
  ): Promise<void> {
    if (!this._client) return;
    const fullKey = `${KEY_PREFIX}:findings:${missionId}:${category}`;
    await this._client.hset(fullKey, key, value);
  }

  async findings_read(
    missionId: string,
    category: string,
    key: string
  ): Promise<string | null> {
    if (!this._client) return null;
    const fullKey = `${KEY_PREFIX}:findings:${missionId}:${category}`;
    return this._client.hget(fullKey, key);
  }

  async findings_read_all(
    missionId: string,
    category: string
  ): Promise<Record<string, string>> {
    if (!this._client) return {};
    const fullKey = `${KEY_PREFIX}:findings:${missionId}:${category}`;
    return this._client.hgetall(fullKey);
  }

  async get_payload_attempt_count(
    missionId: string,
    payloadHash: string
  ): Promise<number> {
    if (!this._client) return 0;
    const fullKey = `${KEY_PREFIX}:payload_attempts:${missionId}`;
    const count = await this._client.hget(fullKey, payloadHash);
    return count ? parseInt(count, 10) : 0;
  }

  async increment_payload_attempt(
    missionId: string,
    payloadHash: string
  ): Promise<number> {
    if (!this._client) return 0;
    const fullKey = `${KEY_PREFIX}:payload_attempts:${missionId}`;
    return this._client.hincrby(fullKey, payloadHash, 1);
  }

  async ping(): Promise<boolean> {
    if (!this._client) return false;
    try {
      const result = await this._client.ping();
      return result === 'PONG';
    } catch {
      return false;
    }
  }

  async store_token(
    missionId: string,
    tokenName: string,
    tokenValue: string
  ): Promise<void> {
    await this.findings_store(missionId, 'tokens', tokenName, tokenValue);
  }

  async get_tokens(missionId: string): Promise<Record<string, string>> {
    return this.findings_read_all(missionId, 'tokens');
  }

  async store_successful_payload(
    missionId: string,
    hash: string,
    payload: string,
    exploitType: string,
    target: string
  ): Promise<void> {
    const fullKey = `${KEY_PREFIX}:findings:${missionId}:successful_payloads`;
    await this._client?.hset(
      fullKey,
      hash,
      JSON.stringify({ payload, exploit_type: exploitType, target, timestamp: new Date().toISOString() })
    );
  }

  async a2a_publish(message: A2AMessage): Promise<void> {
    await this.publish(STREAMS.A2A_MESSAGES, {
      msg_id: message.msg_id,
      sender: message.sender,
      recipient: message.recipient as string,
      type: message.type,
      priority: message.priority,
      payload: JSON.stringify(message.payload),
      timestamp: message.timestamp,
    });
  }

  async a2a_consume(
    group: string,
    consumer: string,
    count: number = 10
  ): Promise<Array<{ id: string; message: A2AMessage }>> {
    const raw = await this.consume(STREAMS.A2A_MESSAGES, group, consumer, count);

    return raw.map(({ id, message }) => ({
      id,
      message: {
        msg_id: message.msg_id,
        sender: message.sender as A2AMessage['sender'],
        recipient: message.recipient as A2AMessage['recipient'],
        type: message.type as A2AMessage['type'],
        priority: message.priority as A2AMessage['priority'],
        payload: JSON.parse(message.payload as string),
        timestamp: message.timestamp,
      } as A2AMessage,
    }));
  }

  private flattenObject(obj: Record<string, string | number | boolean>): string[] {
    const result: string[] = [];
    for (const [key, value] of Object.entries(obj)) {
      result.push(key, String(value));
    }
    return result;
  }

  private unflattenObject(arr: string[]): Record<string, string> {
    const result: Record<string, string> = {};
    for (let i = 0; i < arr.length; i += 2) {
      result[arr[i]] = arr[i + 1];
    }
    return result;
  }
}

export const redisBus = new RedisBus();

File: swarm module/Red_team/core/redis_bus.py
"""
Redis Streams message bus for agent-to-agent (A2A) communication.

Streams:
  - a2a_messages:  Inter-agent task assignments and intelligence reports
  - red_team_events:  Kill chain events for monitoring/logging

Blackboard:
  - HSET/HGET for shared mission state (redteam:blackboard:{mission_id}:*)
"""

from __future__ import annotations

import json
import logging
from typing import Any

import redis.asyncio as aioredis

from core.config import settings

logger = logging.getLogger(__name__)

# Stream names
A2A_STREAM = "a2a_messages"
RED_TEAM_EVENTS = "red_team_events"
DEFENSE_ANALYTICS = "defense_analytics"  # Blue Team -> Red Team bridge


class RedisBus:
    """Async Redis Streams wrapper for agent messaging."""

    def __init__(self, url: str | None = None):
        self._url = url or settings.redis_url
        self._client: aioredis.Redis | None = None

    async def connect(self) -> None:
        """Establish connection to Redis."""
        self._client = aioredis.from_url(
            self._url, decode_responses=True, encoding="utf-8"
        )
        # Verify connectivity
        await self._client.ping()
        logger.info("Redis connected at %s", self._url)

    async def disconnect(self) -> None:
        """Close Redis connection."""
        if self._client:
            await self._client.aclose()
            self._client = None

    @property
    def client(self) -> aioredis.Redis:
        if self._client is None:
            raise RuntimeError("RedisBus not connected — call connect() first")
        return self._client

    # ── Stream Operations ──────────────────────────────────────────

    async def publish(self, stream: str, message: dict[str, Any]) -> str:
        """
        Publish a message to a Redis Stream.
        Returns the stream message ID.
        """
        # Flatten the message — Redis Streams require string field/values
        flat = {k: json.dumps(v) if not isinstance(v, str) else v for k, v in message.items()}
        msg_id = await self.client.xadd(stream, flat)
        logger.debug("Published to %s: %s", stream, msg_id)
        return msg_id

    async def create_consumer_group(
        self, stream: str, group: str, start_id: str = "0"
    ) -> None:
        """Create a consumer group for a stream. Idempotent."""
        try:
            await self.client.xgroup_create(stream, group, id=start_id, mkstream=True)
            logger.info("Created consumer group '%s' on stream '%s'", group, stream)
        except aioredis.ResponseError as e:
            if "BUSYGROUP" in str(e):
                pass  # Group already exists
            else:
                raise

    async def consume(
        self,
        stream: str,
        group: str,
        consumer: str,
        count: int = 10,
        block_ms: int = 5000,
    ) -> list[dict[str, Any]]:
        """
        Read new messages from a consumer group.
        Returns list of dicts with 'msg_id' + original fields.
        """
        results = await self.client.xreadgroup(
            groupname=group,
            consumername=consumer,
            streams={stream: ">"},
            count=count,
            block=block_ms,
        )

        messages = []
        for _stream_name, entries in (results or []):
            for msg_id, fields in entries:
                parsed = {}
                for k, v in fields.items():
                    try:
                        parsed[k] = json.loads(v)
                    except (json.JSONDecodeError, TypeError):
                        parsed[k] = v
                parsed["_msg_id"] = msg_id
                messages.append(parsed)

        return messages

    async def ack(self, stream: str, group: str, *msg_ids: str) -> int:
        """Acknowledge processed messages."""
        return await self.client.xack(stream, group, *msg_ids)

    async def claim_pending(
        self,
        stream: str,
        group: str,
        consumer: str,
        min_idle_ms: int = 60000,
        count: int = 10,
    ) -> list[dict[str, Any]]:
        """
        Claim pending messages that were delivered to other consumers.
        Useful for recovering messages from crashed consumers.
        
        Args:
            stream: Stream name
            group: Consumer group name
            consumer: Consumer name to claim messages for
            min_idle_ms: Minimum idle time before claiming (default 60s)
            count: Max messages to claim
            
        Returns:
            List of claimed messages with 'msg_id' + original fields
        """
        try:
            # Get pending messages info
            pending_info = await self.client.xpending(stream, group)
            if not pending_info or pending_info["pending"] == 0:
                return []
            
            # Get pending message IDs
            pending_entries = await self.client.xpending_range(
                stream, group, min="-", max="+", count=count
            )
            
            if not pending_entries:
                return []
            
            msg_ids = [entry["message_id"] for entry in pending_entries]
            
            # Claim the messages
            claimed = await self.client.xclaim(
                stream, group, consumer, min_idle_time=min_idle_ms, message_ids=msg_ids
            )
            
            messages = []
            for msg_id, fields in claimed:
                parsed = {}
                for k, v in fields.items():
                    try:
                        parsed[k] = json.loads(v)
                    except (json.JSONDecodeError, TypeError):
                        parsed[k] = v
                parsed["_msg_id"] = msg_id
                messages.append(parsed)
            
            return messages
            
        except Exception as e:
            logger.error(f"Failed to claim pending messages: {e}")
            return []

    async def read_stream(
        self, stream: str, count: int = 100, start_id: str = "0"
    ) -> list[dict[str, Any]]:
        """Read messages from a stream (no consumer group — for debugging/monitoring)."""
        results = await self.client.xrange(stream, min=start_id, count=count)
        messages = []
        for msg_id, fields in results:
            parsed = {}
            for k, v in fields.items():
                try:
                    parsed[k] = json.loads(v)
                except (json.JSONDecodeError, TypeError):
                    parsed[k] = v
            parsed["_msg_id"] = msg_id
            messages.append(parsed)
        return messages

    # ── Blackboard Operations ──────────────────────────────────────

    async def blackboard_write(
        self, mission_id: str, key: str, value: Any
    ) -> None:
        """Write a key-value pair to the mission blackboard."""
        bb_key = f"redteam:blackboard:{mission_id}"
        await self.client.hset(bb_key, key, json.dumps(value))
        logger.debug("Blackboard write: %s.%s", mission_id, key)

    async def blackboard_read(self, mission_id: str, key: str) -> Any | None:
        """Read a single key from the mission blackboard."""
        bb_key = f"redteam:blackboard:{mission_id}"
        raw = await self.client.hget(bb_key, key)
        if raw is None:
            return None
        try:
            return json.loads(raw)
        except (json.JSONDecodeError, TypeError):
            return raw

    async def blackboard_read_all(self, mission_id: str) -> dict[str, Any]:
        """Read the entire mission blackboard."""
        bb_key = f"redteam:blackboard:{mission_id}"
        raw = await self.client.hgetall(bb_key)
        parsed = {}
        for k, v in raw.items():
            try:
                parsed[k] = json.loads(v)
            except (json.JSONDecodeError, TypeError):
                parsed[k] = v
        return parsed

    async def blackboard_clear(self, mission_id: str) -> None:
        """Delete the entire mission blackboard."""
        bb_key = f"redteam:blackboard:{mission_id}"
        await self.client.delete(bb_key)

    # ── Defense Analytics (Blue Team Bridge) ───────────────────────

    async def consume_defense_analytics(
        self,
        group: str = "red_team",
        consumer: str = "commander",
        count: int = 10,
        block_ms: int = 5000,
    ) -> list[dict[str, Any]]:
        """
        Consume defensive analytics from the Blue Team.
        
        This allows the Red Team Commander to adapt its strategy based on
        what the Blue Team has detected and blocked.

        Args:
            group: Consumer group name
            consumer: Consumer name
            count: Max messages to read
            block_ms: Block timeout in milliseconds

        Returns:
            List of defense analytics messages
        """
        # Ensure consumer group exists
        await self.create_consumer_group(DEFENSE_ANALYTICS, group)
        
        # Consume messages
        results = await self.client.xreadgroup(
            groupname=group,
            consumername=consumer,
            streams={DEFENSE_ANALYTICS: ">"},
            count=count,
            block=block_ms,
        )

        messages = []
        for _stream_name, entries in (results or []):
            for msg_id, fields in entries:
                parsed = {}
                for k, v in fields.items():
                    try:
                        parsed[k] = json.loads(v)
                    except (json.JSONDecodeError, TypeError):
                        parsed[k] = v
                parsed["_msg_id"] = msg_id
                messages.append(parsed)

        return messages

    async def get_latest_defense_intel(
        self,
        count: int = 50,
    ) -> list[dict[str, Any]]:
        """
        Get recent defense analytics (no consumer group - for polling).

        Args:
            count: Maximum number of messages to retrieve

        Returns:
            List of recent defense analytics
        """
        results = await self.client.xrevrange(
            DEFENSE_ANALYTICS,
            "+",
            "-",
            count=count,
        )

        messages = []
        for msg_id, fields in results:
            parsed = {}
            for k, v in fields.items():
                try:
                    parsed[k] = json.loads(v)
                except (json.JSONDecodeError, TypeError):
                    parsed[k] = v
            parsed["_msg_id"] = msg_id
            messages.append(parsed)

        return messages

    # ── Shared Findings Store (PentAGI v4.0 Memory System) ────────

    async def findings_store(
        self, mission_id: str, category: str, key: str, value: Any
    ) -> None:
        """
        Store a finding for cross-agent propagation.

        Categories: tokens, credentials, successful_payloads, endpoints, owasp_successes
        Example: findings_store("m1", "tokens", "admin_jwt", "eyJ...")
        """
        findings_key = f"redteam:findings:{mission_id}:{category}"
        await self.client.hset(findings_key, key, json.dumps(value) if not isinstance(value, str) else value)
        logger.info("📦 Findings store: %s/%s.%s", mission_id, category, key)

    async def findings_read(self, mission_id: str, category: str) -> dict[str, Any]:
        """Read all findings in a category."""
        findings_key = f"redteam:findings:{mission_id}:{category}"
        raw = await self.client.hgetall(findings_key)
        parsed = {}
        for k, v in raw.items():
            try:
                parsed[k] = json.loads(v)
            except (json.JSONDecodeError, TypeError):
                parsed[k] = v
        return parsed

    async def findings_read_all(self, mission_id: str) -> dict[str, dict[str, Any]]:
        """Read all findings across all categories."""
        result = {}
        for category in ("tokens", "credentials", "successful_payloads", "endpoints", "owasp_successes"):
            data = await self.findings_read(mission_id, category)
            if data:
                result[category] = data
        return result

    async def log_parse_failure(
        self, mission_id: str, agent: str, raw_text: str
    ) -> None:
        """Log a JSON parse failure for critic analysis."""
        failure_key = f"redteam:parse_failures:{mission_id}"
        entry = json.dumps({"agent": agent, "text": raw_text[:500]})
        await self.client.rpush(failure_key, entry)
        logger.debug("Parse failure logged for critic: %s", agent)

    async def get_payload_attempt_count(
        self, mission_id: str, payload_hash: str
    ) -> int:
        """Get how many times a payload has been attempted (max 2 retries)."""
        key = f"redteam:payload_attempts:{mission_id}"
        count = await self.client.hget(key, payload_hash)
        return int(count) if count else 0

    async def increment_payload_attempt(
        self, mission_id: str, payload_hash: str
    ) -> int:
        """Increment and return attempt count for a payload."""
        key = f"redteam:payload_attempts:{mission_id}"
        return await self.client.hincrby(key, payload_hash, 1)

    # ── Health Check ───────────────────────────────────────────────

    async def ping(self) -> bool:
        """Check Redis connectivity."""
        try:
            return await self.client.ping()
        except Exception:
            return False


# Default singleton
redis_bus = RedisBus()
